dashboard/app/models/lesson_group.rb

```diff
@@ -36,6 +38,91 @@ class LessonGroup < ApplicationRecord
     display_name
   )
 
+  # Finds or creates Lesson Groups with the correct position.
+  # In addition it check for 3 things:
+  # 1. That all the lesson groups specified by the editor have a key and
+  # display name.
+  # 2. If the lesson group key is an existing key that the given display name
+  # for that key matches the already saved display name
+  # 3. PLC courses use certain lesson group keys for module types. We reserve those
+  # keys so they can only map to the display_name for their PLC purpose
+  def self.add_lesson_groups(raw_lesson_groups, script, new_suffix, editor_experiment)
+    lockable_count = 0
+    non_lockable_count = 0
+    lesson_position = 0
+    position = 0
+    chapter = 0
+
+    raw_lesson_groups&.map do |raw_lesson_group|
+      if raw_lesson_group[:key].nil?
+        lesson_group = LessonGroup.find_or_create_by(
+          key: '',
+          script: script,
+          user_facing: false,
+          position: 1
+        )
+      else
+        LessonGroup.prevent_changing_plc_display_name(raw_lesson_group)
+        LessonGroup.prevent_blank_display_name(raw_lesson_group)
+
+        new_lesson_group = false
+
+        lesson_group = LessonGroup.find_or_create_by(
+          key: raw_lesson_group[:key],
+          script: script,
+          user_facing: true
+        ) do
+          # if you got in here, this is a new lesson_group
+          new_lesson_group = true
+        end
+
+        LessonGroup.prevent_changing_display_name_for_existing_key(new_lesson_group, raw_lesson_group, lesson_group)
+
+        lesson_group.assign_attributes(position: position += 1, properties: {display_name: raw_lesson_group[:display_name]})
+        lesson_group.save! if lesson_group.changed?
+      end
+
+      LessonGroup.prevent_lesson_group_with_no_lessons(lesson_group, raw_lesson_group[:lessons].length)
+
+      new_lessons = Lesson.add_lessons(script, lesson_group, raw_lesson_group[:lessons], lockable_count, non_lockable_count, lesson_position, chapter, new_suffix, editor_experiment)
+      lesson_group.lessons = new_lessons
+      lesson_group.save!
+
+      lesson_position += lesson_group.lessons.length
```

Comment:
This pattern of passing in all the position counters (lockable_count, non_lockable_count, lesson_position, position, chapter) into the lower-level methods, and then afterwards figuring out how to update them in the higher-level methods, seems a bit redundant and fragile to me. The lower-level methods already incremented the values locally so ideally we shouldn't repeat that work somewhere else.

One option I can think of is to create a single "Counters" Struct to hold these five values, and to pass that object into each of the lower-level methods instead of passing them all separately. Then, when the lower-level methods increment the counters, those changes will automatically be reflected in the higher-level methods already, since everything is sharing that one "Counters" object.

This would also have the side benefit of shortening the parameter lists for some of these methods and making them more readable.

Reply:
I will follow up with another PR to do this but I want to get these changes in first.

Reply:
https://codedotorg.atlassian.net/browse/PLAT-252